### PR TITLE
18951 Migrate from deprecated ui-select2 to ui-select in timelines view

### PR DESCRIPTION
**Pull request by myabc**
:traffic_light: This PR allows us to finally get rid of ui-select2
- Add basic spec coverage.

https://community.openproject.org/work_packages/18951

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://myabc-feature-18951-ui-select-timelines.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://myabc-feature-18951-ui-select-timelines.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

File: spec/helpers/timelines_helper_spec.rb [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See doc/COPYRIGHT.rdoc for more details.
+#++
+
+require 'spec_helper'
+
+describe TimelinesHelper, type: :helper do
+  describe '#push_visible_timelines' do
+    let(:project)   { FactoryGirl.create(:project) }
```
Comment (myabc) [inline spec/helpers/timelines_helper_spec.rb:34]: @houndci since when has this bothered you?

**Comment by NobodysNightmare** [comment]
I am missing the dropdown button on the right hand side:

combobox

Currently it does not look like a combobox-ish control, but like a textbox.

**Comment by NobodysNightmare** [comment]
The combobox is trolling me: [click for awesome GIF](https://gfycat.com/LittleEminentCarpenterant)

When trying to filter the combobox it highlights the entry I am looking for, but it does not actually select it, neither does it remove entries that are not matching my selection...

Pressing enter loads (as expected) the entry I did not want, but whcih was selected.

**Comment by myabc** [comment]
@NobodysNightmare is this happening in other comboboxes? (e.g. work package details view)

File: app/helpers/timelines_helper.rb [hunk 2]
```diff
@@ -307,25 +307,21 @@ def internationalized_columns_select(collection)
...
-  def visible_timeline_paths(_visible_timelines = [])
-    @visible_timelines.inject({}) do |timeline_paths, timeline|
-      timeline_paths.merge(timeline.id => { path: project_timeline_path(@project, timeline) })
-    end
-  end
-
-  def push_visible_timeline_paths(visible_timelines)
-    gon.timelines = visible_timeline_paths visible_timelines
+  def push_visible_timelines(visible_timelines, target = gon)
+    target.timelines = visible_timelines.map { |timeline|
```
Comment (NobodysNightmare) [inline app/helpers/timelines_helper.rb:311]: Assuming that you are doing a refactoring in this file:

Did you notice, that you changed the behaviour here?
Previously the `_visible_timelines` parameter was ignored and the instance variable `@visible_timelines` was taken instead.

Your new code really cares for the parameter it receives.

While this sounds like a sensible move, I want to make sure it was an intended one...

**Comment by NobodysNightmare** [comment]
@myabc Now that you say it: Yes it is the same for the details view controls.

I did never realize it there, because upon opening them for the first time they have that little edit icon on the right side, so I simply click them and if a dropdown happens to pop up I am happy.

After selecting an entry from the dropdown the edit icon is not there and IMHO the dropdown button is missing (i.e. the button does not need to be visible as long as I have the edit icon)

inplace

**Comment by NobodysNightmare** [comment]
That is all my own opinion though, don't know what the style guide says about that...

**Comment by NobodysNightmare** [comment]
Regarding the [filtering issue](https://gfycat.com/LittleEminentCarpenterant):

This is only happening for the timeline select combobox, details pane works fine...

**Comment by myabc** [comment]
not sure if this was the same issue @vcostin was working on? (it's been a while since I rebased)

**Comment by myabc** [comment]
> Regarding the filtering issue:
> 
> This is only happening for the timeline select combobox, details pane works fine...

@NobodysNightmare Thanks for awesome GIFs.  6eb6612 should resolve this issue.

**Comment by myabc** [comment]
@NobodysNightmare mind manually merging this into `release/4.1`? – it'd save me a few minutes and keep the excellent comments in tact :wink:

**Comment by NobodysNightmare** [comment]
of course... let me just verify your fix :wink:

**Comment by NobodysNightmare** [comment]
It now works as intended, but I am not sure if I am seeing the right thing.

Can you verify that this is what I should see:

right-thing

I am in doubt because:
- I already needed to delete `tmp` and `.sass-cache` because otherwise buttons looked wrong
- I expected the select element to look like the one I posted previously, but it doesn't...

**Comment by myabc** [comment]
Ignore the aesthetics of the buttons and UI-select element. The branch
hasn't been rebased in some time and doesn't include the Foundation upgrade
or UI-select styling commits.

**Comment by NobodysNightmare** [comment]
Merged it into `dev` and merged it into `release/4.1` afterwards -.-

One extra commit ^^

**Pull request merged**